src/Microsoft.VisualStudio.Composition/CompositionConfiguration.cs

```diff
@@ -150,27 +142,54 @@ public static CompositionConfiguration Create(ComposableCatalog catalog)
             // Detect loops of all non-shared parts.
             errors.AddRange(FindLoops(parts));
 
+            // If errors are found, re-validate the salvaged parts in case there are parts whose dependencies are affected by the initial errors
             if (errors.Count > 0)
             {
-                var invalidParts = ImmutableHashSet.CreateRange(errors.SelectMany(error => error.Parts).Select(p => p.Definition));
-                if (invalidParts.IsEmpty)
+                // Set salvaged parts to current catalog in case there are no errors
+                var salvagedParts = parts;
+                var salvagedPartDefinitions = catalog.Parts;
+
+                List<ComposedPartDiagnostic> previousErrors = errors;
+                ImmutableQueue<IReadOnlyCollection<ComposedPartDiagnostic>> compositionErrors = ImmutableQueue<IReadOnlyCollection<ComposedPartDiagnostic>>.Empty;
+
+                // While we still find errors we validate the exports so that we remove all dependency failures
+                while (previousErrors.Count > 0)
                 {
-                    // If we can't identify the faulty parts but we still have errors, we have to just throw.
-                    throw new CompositionFailedException(Strings.FailStableComposition, ImmutableStack.Create<IReadOnlyCollection<ComposedPartDiagnostic>>(errors));
+                    compositionErrors = compositionErrors.Enqueue(previousErrors);
+
+                    // Get the salvaged parts
+                    var invalidParts = previousErrors.SelectMany(error => error.Parts);
```

Comment:
`invalidParts` is implicitly typed as `IEnumerable<T>`, and is enumerated *twice* below. That means you're executing the `SelectMany` twice. Also given that `IEnumerable<T>` is not a concrete collection type, each enumeration allocates an `IEnumerator<T>`. Can you add `.ToList()` to the end of this expression so that `SelectMany` executes only once and you end with a `List<T>` instead?